File: Iterator Design Pattern/C++/amazon_inventory.cpp
#include <iostream>

using namespace std;

// Product class representing individual products
class Product {
    private:
        string name;
        double price;

    public:
        Product(const string& name, const double& price) : name(name), price(price) {}

        const string& getName() const {
            return name;
        }

        double getPrice() const {
            return price;
        }
};

// Iterator interface
class Iterator {
    public:
        virtual ~Iterator() {}
        virtual Product* first() = 0;
        virtual Product* next() = 0;
        virtual bool hasNext() = 0;
};

// Concrete iterator for the product collection
class ProductIterator : public Iterator {
    private:

        vector<Product*>& products;
        size_t current;

    public:

        ProductIterator(vector<Product*>& products) : products(products), current(0) {}

        Product* first() {
            if (products.empty()) {
                return nullptr;
            }
            current = 0;
            return products[current];
        }

        Product* next() {
            if (hasNext()) {
                return products[++current];
            }
            return nullptr;
        }

        bool hasNext() {
            return current < products.size();
        }
};

// Aggregate class that stores products and provides an iterator
class Inventory {
    private:
        vector<Product*> products;
    public:
        void addProduct(Product* product) {
            products.push_back(product);
        }

        Iterator* createIterator() {
            return new ProductIterator(products);
        }
};


int main() {

    // Create some products
    Product product1("Laptop", 99999.99);
    Product product2("Smartphone", 49999.99);
    Product product3("Headphones", 7999.99);

    // Create an inventory and add products
    Inventory inventory;
    inventory.addProduct(&product1);
    inventory.addProduct(&product2);
    inventory.addProduct(&product3);

    // Create an iterator and iterate over the products
    Iterator* iterator = inventory.createIterator();
    Product* currentProduct = iterator->first();

    while (currentProduct) {
        cout << "Product: " << currentProduct->getName() << 
                        ", Price: $" << currentProduct->getPrice() << std::endl;
        currentProduct = iterator->next();
    }

    delete iterator;

    return 0;
}

File: Iterator Design Pattern/JAVA/AmazonInventory.java
import java.util.*;

// Product class representing individual products
class Product {
    private String name;
    private double price;

    public Product(String name, double price) {
        this.name = name;
        this.price = price;
    }

    public String getName() {
        return name;
    }

    public double getPrice() {
        return price;
    }
}

// Iterator interface
interface Iterator {
    Product first();
    Product next();
    boolean hasNext();
}

// Concrete iterator for the product collection
class ProductIterator implements Iterator {
    private List<Product> products;
    private int current;

    public ProductIterator(List<Product> products) {
        this.products = products;
        this.current = 0;
    }

    public Product first() {
        if (products.isEmpty()) {
            return null;
        }
        current = 0;
        return products.get(current);
    }

    public Product next() {
        if (hasNext()) {
            return products.get(++current);
        }
        return null;
    }

    public boolean hasNext() {
        return current < products.size() - 1;
    }
}

// Aggregate class that stores products and provides an iterator
class Inventory {
    private List<Product> products = new ArrayList<>();

    public void addProduct(Product product) {
        products.add(product);
    }

    public Iterator createIterator() {
        return new ProductIterator(products);
    }
}

public class AmazonInventory {
    public static void main(String[] args) {
        // Create some products
        Product product1 = new Product("Laptop", 99999.99);
        Product product2 = new Product("Smartphone", 49999.99);
        Product product3 = new Product("Headphones", 7999.99);

        // Create an inventory and add products
        Inventory inventory = new Inventory();
        inventory.addProduct(product1);
        inventory.addProduct(product2);
        inventory.addProduct(product3);

        // Create an iterator and iterate over the products
        Iterator iterator = inventory.createIterator();
        Product currentProduct = iterator.first();

        while (currentProduct != null) {
            System.out.println("Product: " + currentProduct.getName() + ", Price: $" + currentProduct.getPrice());
            currentProduct = iterator.next();
        }
    }
}
